File: src/main/java/masonry/api/Transition.java
package masonry.api;

import net.scriptgate.common.Point;

public interface Transition {

    int toX();

    int toY();

    float getPercentage();

    Point getLocationAt(double elapsedTime);

    boolean isCompleted();

    static Transition none(int x, int y) {
        return new Transition() {

            @Override
            public Point getLocationAt(double elapsedTime) {
                return new Point(x, y);
            }

            @Override
            public int toX() {
                return x;
            }

            @Override
            public int toY() {
                return y;
            }

            @Override
            public float getPercentage() {
                return 1;
            }

            @Override
            public boolean isCompleted() {
                return true;
            }
        };
    }
}

File: src/main/kotlin/net/scriptgate/masonry/api/Transition.kt
package net.scriptgate.masonry.api

import net.scriptgate.common.Point

interface Transition {

    val percentage: Float

    val isCompleted: Boolean

    fun toX(): Int

    fun toY(): Int

    fun getLocationAt(elapsedTime: Double): Point

    companion object {

        fun none(x: Int, y: Int): Transition {
            return object : Transition {

                override val percentage: Float
                    get() = 1f

                override val isCompleted: Boolean
                    get() = true

                override fun getLocationAt(elapsedTime: Double): Point {
                    return Point(x, y)
                }

                override fun toX(): Int {
                    return x
                }

                override fun toY(): Int {
                    return y
                }
            }
        }
    }
}
